src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/rulemanager/OfBatchHolder.java

```diff
@@ -165,153 +171,84 @@ public OfBatchHolder(IOFSwitchService iofSwitchService, MessageContext messageCo
-        OFFactory factory = iofSwitchService.getSwitch(dpId).getOFFactory();
-        OFMessage message = OfMeterConverter.INSTANCE.convertDeleteMeterCommand(data, factory);
-        xidMapping.put(message.getXid(), data.getUuid());
-        metersMap.put(data.getMeterId(), data);
-        BatchData batchData = BatchData.builder()
-                .meter(true)
-                .message(message)
-                .origin(data)
-                .presenceBeVerified(false)
-                .build();
-        addBatchData(data.getUuid(), batchData, data.getDependsOn());
+        addSpeakerData(data, switchId, OfMeterConverter.INSTANCE::convertDeleteMeterCommand, metersMap,
+                MeterSpeakerData::getMeterId, false);
     }
 
     @Override
     public void addInstallGroup(GroupSpeakerData data, SwitchId switchId) {
-        DatapathId dpId = DatapathId.of(switchId.toLong());
-        OFFactory factory = iofSwitchService.getSwitch(dpId).getOFFactory();
-        OFMessage message = OfGroupConverter.INSTANCE.convertInstallGroupCommand(data, factory);
-        xidMapping.put(message.getXid(), data.getUuid());
-        groupsMap.put(data.getGroupId(), data);
-        BatchData batchData = BatchData.builder()
-                .group(true)
-                .message(message)
-                .origin(data)
-                .presenceBeVerified(true)
-                .build();
-        addBatchData(data.getUuid(), batchData, data.getDependsOn());
+        addSpeakerData(data, switchId, OfGroupConverter.INSTANCE::convertInstallGroupCommand, groupsMap,
+                GroupSpeakerData::getGroupId, true);
     }
 
     @Override
     public void addModifyGroup(GroupSpeakerData data, SwitchId switchId) {
-        DatapathId dpId = DatapathId.of(switchId.toLong());
-        OFFactory factory = iofSwitchService.getSwitch(dpId).getOFFactory();
-        OFMessage message = OfGroupConverter.INSTANCE.convertModifyGroupCommand(data, factory);
-        xidMapping.put(message.getXid(), data.getUuid());
-        BatchData batchData = BatchData.builder()
-                .group(true)
-                .message(message)
-                .origin(data)
-                .presenceBeVerified(true)
-                .build();
-        commandMap.put(data.getUuid(), batchData);
-        addBatchData(data.getUuid(), batchData, data.getDependsOn());
+        addSpeakerData(data, switchId, OfGroupConverter.INSTANCE::convertModifyGroupCommand, groupsMap,
+                GroupSpeakerData::getGroupId, true);
     }
 
     @Override
     public void addDeleteGroup(GroupSpeakerData data, SwitchId switchId) {
+        addSpeakerData(data, switchId, OfGroupConverter.INSTANCE::convertDeleteGroupCommand, groupsMap,
+                GroupSpeakerData::getGroupId, false);
+    }
+
+    private <S extends SpeakerData, K> void addSpeakerData(S data, SwitchId switchId,
+                                                           BiFunction<S, OFFactory, OFMessage> converter,
```

Comment:
Converter will be not called in case of switch unavailable. Different converter should be used for install/modify/delete commands so we still need to pass this param here. I don't see the benefits to store converter function itself and call it only before execution.